### PR TITLE
feat: expose logs in OutputChannel, log info on startup

### PR DESCRIPTION
**Pull request by justinmk**
- Introduce `Logger.info()` and "info" log-level.
- Cleanup settings descriptions.
- Create outputChannel as a `{ log: true }` channel for enhanced rendering and features.
- Allow the logger to log to the vscode Output channel. 
    - But disable it for noisy (and looping) sources like `CursorManager` and `DocumentChangeManager`.

image

### CODE REVIEW

File: src/logger.ts [hunk 1]
```diff
@@ -3,20 +3,35 @@ import fs from "fs";
...
+    info = 3,
+    debug = 4,
 }
 
 export interface ILogger {
     debug(...args: any[]): void;
+    info(...args: any[]): void;
     warn(...args: any[]): void;
     error(...args: any[]): void;
+    /**
```
Comment (theol0403) [inline src/logger.ts:24]: How does this work with the old `debug`/`warn` methods? Should we stick to those (and submit the uri to them), or remove them in favour of `log`? Which one should you use?

**Review by theol0403 — COMMENTED** [review]
Very awesome!

> Disable it for noisy (and looping) sources like CursorManager and DocumentChangeManager.

Won't this cause confusion when debugging `CursorManager` and seeing no logs come through the output?

**Comment by justinmk** [comment]
> Won't this cause confusion when debugging `CursorManager` and seeing no logs come through the output?

It's only "disabled" for the output channel (to avoid infinite loop). If one needs to debug cursor movement they can use the dev tools console or log-file.

**Comment by theol0403** [comment]
> It's only disabled for the output channel (to avoid infinite loop). If one needs to debug cursor movement they can use the dev tools console or log-file.

Just make sure this is documented because I see people trying to use the output channel for all debugging without knowing the dev-tools exists.

File: src/logger.ts (hunk 1 above)
Comment (justinmk) [reply src/logger.ts:24]: Adding `uri` to the existing debug/warn/error methods would be very high-churn for a somewhat limited use-case. (Possibly could be done with typescript "overload" signature, but also not sure that's worth the effort).

Comment (justinmk) [reply src/logger.ts:24]: Is the docstring here not sufficient?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -35,6 +35,14 @@ npx vsce package -o vscode-neovim.vsix
 1. Open the repo in VSCode.
 2. Go to debug view and click `Run Extension` (F5).
 
+### Logging
+
+You can observe the extension logs via the `vscode-neovim` Output channel or from the dev tools
+console (run the `Developer: Toggle Developer Tools` vscode command to see the console).
+
+Note: some messages are not logged to the Output channel, to avoid infinite loop. This is decided by
```
Comment (justinmk) [inline CONTRIBUTING.md:43]: added doc

File: src/main_controller.ts [hunk 3]
```diff
@@ -57,14 +57,14 @@ export class MainController implements vscode.Disposable {
...
+    public messagesManager!: MessagesManager;
     public viewportManager!: ViewportManager;
 
     public constructor(private extContext: ExtensionContext) {}
 
-    public async init(): Promise<void> {
+    public async init(outputChannel: vscode.LogOutputChannel): Promise<void> {
         const [cmd, args] = this.buildSpawnArgs();
-        logger.debug(`Spawning nvim: ${cmd} ${args.join(" ")}`);
+        logger.info(`starting nvim: ${cmd} ${args.join(" ")}`);
```
Comment (justinmk) [inline src/main_controller.ts:67]: Log Nvim start-command at startup. Will add more info logs in the future (e.g. the result of `stdpath('config')`)

File: src/logger.ts (hunk 1 above)
Comment (theol0403) [reply src/logger.ts:24]: If I were a new contributor adding a new feature, I would be unsure whether to use `logger.warn` or `logger.log(uri, LogLevel.warn)`

Comment (justinmk) [reply src/logger.ts:24]: I understand. But in this case I can't think of a general guidance. New contributors sometimes need to figure things out. If they choose the "wrong" one, the cost is negligible, in this cas.

**Review by theol0403 — APPROVED** [review]

**Pull request merged**